### PR TITLE
Add new base validator and base constraint classes

### PR DESCRIPTION
**Pull request by drmull**
Add new base classes to validator and constraint that take additional arguments, such as schema, path etc. This allows constraints to use non-primitive validators and resolves issue #133 

My initial intention was to pull out the logic for the non-primitive validators from the schema file and put them in each respective validator class but then I figured it would be a bit too drastic..

### CODE REVIEW

File: yamale/validators/base.py [hunk 1]
```diff
@@ -30,37 +30,32 @@ def tag(self):
...
-        valid = self._is_valid(value)
-        if not valid:
-            errors.append(self.fail(value))
+        # Optional field with optional value? Who cares.
+        if (value is None and
+                not self.is_required and
+                self._value_can_be_none):
+            return []
+
+        errors = self._validate(value, schema, path, strict)
```
Comment (drmull) [inline yamale/validators/base.py:52]: I forgot to change the name of the _validate function, I will fix it

File: yamale/validators/constraints.py [hunk 2]
```diff
@@ -107,20 +119,22 @@ def _fail(self, value):
...
-class Key(Constraint):
+class Key(BaseConstraint):
     keywords = {'key': Validator}
     fail = 'Key error - %s'
 
-    def _is_valid(self, value):
+    def _is_valid(self, value, schema, path, strict):
         for k in value.keys():
-            if self.key.validate(k) != []:
+            errors = schema._validate(self.key, k, DataPath(), strict)
```
Comment (drmull) [inline yamale/validators/constraints.py:128]: I forgot to change the name of the _validate function, I will fix it

**Comment by mildebrandt** [comment]
Welcome back @drmull :) 

FYI: This may take me a while to get to, it's a busy time over here. Sorry.

**Comment by drmull** [comment]
> Welcome back @drmull :)

Thanks!

> 
> FYI: This may take me a while to get to, it's a busy time over here. Sorry.

No problem, I am in no rush :)

**Comment by mildebrandt** [comment]
Sorry for the long delay. 

I grabbed the example from #133, but I couldn't get your code to work with it. 

**data**:
```
maptest: 
  1: xpto 
  2: qwerty 
  error: wrong
```

**schema**:
```
maptest: map(include('value'), key=include('key')) 
--- 
value: str()
key: int()
```

This combination should fail validation.

**Comment by drmull** [comment]
Hm, strange. I updated the test case to match the example but I get:

`maptest: Key error - 'error' is not a int.`

Do you use the PyYAML parser? Maybe the order of the elements can differ with different parsers and trigger the problem.

I will take a closer look!

**Comment by mildebrandt** [comment]
Yeah, I'm using PyYaml 5.3.1.

However, I'm beginning to think this isn't something we should support. Allowing for `include` in the `key` constraint only makes sense when processing complex keys, for example:

```
? {one: 1, two: 2}
: [ 2001-01-01, 2002-02-02 ]
```

Both PyYAML and ruamel fail to load such a document, but it's valid yaml. Instead, I think the correct course of action is to raise an error saying something like: 

```
The include validator is not valid for the key constraint.
```

Other validators can be used to constrain the `key` and still support all yaml documents that PyYAML and ruamel can load.

**Comment by drmull** [comment]
Yeah, I can agree with that. I will close this PR and we can reopen it if ruamel implements complex keys in the future :grin:

**Comment by mildebrandt** [comment]
Cool, and thanks for your willingness to always jump in!

**Pull request closed without merging**